### PR TITLE
Klee tests cover ring buffer and string

### PR DESCRIPTION
**Pull request by LinerSu**
klee's running time on Ring Buffer:
```
    Start 273: klee_ring_buffer_clean_up_test
1/6 Test #273: klee_ring_buffer_clean_up_test ..............   Passed   11.32 sec
    Start 275: klee_ring_buffer_init_test
2/6 Test #275: klee_ring_buffer_init_test ..................   Passed   12.24 sec
    Start 295: klee_ring_buffer_release_test
3/6 Test #295: klee_ring_buffer_release_test ...............   Passed   46.40 sec
    Start 297: klee_ring_buffer_acquire_test
4/6 Test #297: klee_ring_buffer_acquire_test ...............   Passed   50.66 sec
    Start 299: klee_ring_buffer_buf_belongs_to_pool_test
5/6 Test #299: klee_ring_buffer_buf_belongs_to_pool_test ...   Passed   43.59 sec
    Start 301: klee_ring_buffer_acquire_up_to_test
6/6 Test #301: klee_ring_buffer_acquire_up_to_test .........   Passed   61.42 sec
```

klee's running time on String:
```
      Start 231: klee_string_bytes_test
 1/15 Test #231: klee_string_bytes_test ........................   Passed   21.86 sec
      Start 233: klee_string_compare_test
 2/15 Test #233: klee_string_compare_test ......................   Passed   97.90 sec
      Start 235: klee_string_eq_test
 3/15 Test #235: klee_string_eq_test ...........................   Passed   42.80 sec
      Start 237: klee_string_eq_byte_buf_test
 4/15 Test #237: klee_string_eq_byte_buf_test ..................   Passed   92.60 sec
      Start 239: klee_string_eq_byte_buf_ignore_case_test
 5/15 Test #239: klee_string_eq_byte_buf_ignore_case_test ......   Passed  105.94 sec
      Start 241: klee_string_eq_byte_cursor_test
 6/15 Test #241: klee_string_eq_byte_cursor_test ...............   Passed   19.45 sec
      Start 243: klee_string_eq_byte_cursor_ignore_case_test
 7/15 Test #243: klee_string_eq_byte_cursor_ignore_case_test ...   Passed   24.00 sec
      Start 245: klee_string_eq_c_str_test
 8/15 Test #245: klee_string_eq_c_str_test .....................   Passed  127.40 sec
      Start 247: klee_string_eq_c_str_ignore_case_test
 9/15 Test #247: klee_string_eq_c_str_ignore_case_test  ...   Passed  362.78 sec
      Start 249: klee_string_eq_ignore_case_test
10/15 Test #249: klee_string_eq_ignore_case_test ...............   Passed  104.85 sec
      Start 251: klee_string_new_from_array_test
11/15 Test #251: klee_string_new_from_array_test  ......................   Passed  324.32 sec
      Start 253: klee_string_new_from_c_str_test
12/15 Test #253: klee_string_new_from_c_str_test ...............   Passed   15.08 sec
      Start 255: klee_string_new_from_string_test
13/15 Test #255: klee_string_new_from_string_test ..............  Passed  380.23 sec
      Start 257: klee_string_destroy_test
14/15 Test #257: klee_string_destroy_test ......................   Passed   10.37 sec
      Start 259: klee_string_destroy_secure_test
15/15 Test #259: klee_string_destroy_secure_test ...............   Passed   10.02 sec
```

### CODE REVIEW

File: seahorn/jobs/string_new_from_string/aws_string_new_from_string_harness.c [hunk 1]
```diff
@@ -12,7 +12,7 @@
 
 int main() {
     /* parameters */
-    struct aws_string *source = ensure_string_is_allocated_bounded_length(SIZE_MAX - 1 - sizeof(struct aws_string));
+    struct aws_string *source = ensure_string_is_allocated_nondet_length();
```
Comment (LinerSu) [inline seahorn/jobs/string_new_from_string/aws_string_new_from_string_harness.c:15]: call `ensure_string_is_allocated_nondet_length()` is equilvance to call `ensure_string_is_allocated_bounded_length(SIZE_MAX - 1 - sizeof(struct aws_string))` for seahorn.
The definition is here:
https://github.com/seahorn/verify-c-common/blob/6bd27abc24af7140e2e2270afd8ce03fef1a9de0/seahorn/lib/string_helper.c#L29-L34

File: seahorn/jobs/string_destroy_secure/aws_string_destroy_secure_harness.c [hunk 2]
```diff
@@ -19,6 +19,11 @@ int main(void) {
   bool nondet_parameter = nd_bool();
   aws_string_destroy_secure(nondet_parameter ? str : NULL);
   if (nondet_parameter) {
+    #ifdef __KLEE__
+    // str (even bytes) becomes a dangling pointer used after free
+    // klee does not allow to access memory after free.
+    bytes = NULL;
+    #endif
```
Comment (agurfinkel) [inline seahorn/jobs/string_destroy_secure/aws_string_destroy_secure_harness.c:26]: `assert_all_zeroes` should be in the `#else` branch of the `#ifdef`. This way you trick the assertion to pass, but for KLEE the assertion cannot be checked.

File: seahorn/lib/klee_byte_buf_helper.c [hunk 3]
```diff
@@ -55,9 +55,13 @@ bool aws_byte_cursor_is_bounded(
 
 void ensure_byte_buf_has_allocated_buffer_member(
     struct aws_byte_buf *const buf) {
-    //TODO - figure out allocator issues
     buf->allocator = sea_allocator();
-    buf->buffer = bounded_malloc(sizeof(*(buf->buffer)) * sea_max_buffer_size());
+    if (buf) {
```
Comment (agurfinkel) [inline seahorn/lib/klee_byte_buf_helper.c:59]: `if` is not needed. If the `if` fails, then `buf` is not allocated and previous line is a segfault.
If previous line succeeds, then `if` succeeds.

**Review by agurfinkel — CHANGES_REQUESTED** [review]
A few little comments, but otherwise good to go

**Review by agurfinkel — APPROVED** [review]

**Pull request merged**